### DISCUSSION TITLE
Integrating Vue JS Components with Websight

### DISCUSSION THREAD

**Discussion by Harikarthyk** [websight-io/starter#78] [Q&A]
Hey! 👋
We attempted to integrate WebSight with Vue.

So, in the page folder under the resources, we created the Vue app instance with primevue (for styling) in the body.html.

So the Vue components are rendering. But when we edit the Vue components after clicking submit button in the Edit dialog box the Vue instance or the javascript that is responsible for Vue components, it is not rendering. But I see my changes coming as a response from rootcontainer.html. I believe the Vue thing is not persisting after an edit submission, but we can see the changes after a refresh.

Is there any way or idea to implement Javascript (Vue) in WebSight so that the Vue thing remains after editing?

**Comment by paweljozwicki** [comment]
HI!
Can I ask you to provide some more details? Maybe you can share some screenshots and code samples to ensure we will recreate your issue?

**Reply by Harikarthyk** [reply]
1. So in the page folder.
   body-content.html is under the div tag which has the id "app", which is basically the Vue component mounted.

![Screenshot 2022-12-05 at 4 36 21 PM](https://user-images.githubusercontent.com/54505967/205622463-095635e4-c385-4798-8967-5db602daf7be.png)


2. So in the head.libs, all the Vue, and prime Vue Cdn links are added.

![Screenshot 2022-12-05 at 4 44 18 PM](https://user-images.githubusercontent.com/54505967/205624005-33fed4b1-26cc-4d8a-abae-31b62cb2e5ad.png)


3. So in the cta.html (which is basically the component that is created similar to howlite). in this p-button tag is used which is defined in the body-end.html. (please check the 4th screenshot for the same.)

![MicrosoftTeams-image](https://user-images.githubusercontent.com/54505967/205624327-84495cdb-d604-4f26-b895-534251145a15.png)

4. In the body-end.html, where the Vue instance is added.

![MicrosoftTeams-image (2)](https://user-images.githubusercontent.com/54505967/205625270-b323b558-cb01-4672-bbfe-b48649f9f18a.png)

**Reply by Shivam-apna** [reply]
@paweljozwicki  You can also suggest us a better way to integrate prime-vue and vue in websight as I I didn't find any specific thing in documentation.

**Reply by Harikarthyk** [reply]
Initially, things are working as expected.

![Screenshot 2022-12-05 at 5 22 44 PM](https://user-images.githubusercontent.com/54505967/205632328-121c0d33-bce6-4c7a-bb0c-001d38f3dc0e.png)

After editing this component, Vue JS is not loaded. I tried with a normal button tag with CSS, and it is working, but I am facing issues when I use vue components.

![Screenshot 2022-12-05 at 5 22 58 PM](https://user-images.githubusercontent.com/54505967/205631913-221de04d-8cdf-4056-8b16-e6dfcc5d848a.png)


After refreshing got the required output.
![Screenshot 2022-12-05 at 5 30 28 PM](https://user-images.githubusercontent.com/54505967/205632401-fabda3d4-83c5-4efc-8cff-d2ae6f0e0ba3.png)

**Reply by paweljozwicki** [reply]
Thanks for all the details. 
We have already found the reason for your problems. After saving changes in the dialog, the page editor reloads the page content, but it is done by changing the main container HTML instead of reloading the whole page. That's why the button label is changed, but the Vue app is not re-rendered. 

Now we are discussing some possible solutions. We will try to recreate your environment and provide something.

**Comment by michalcukierman** [comment]
Maybe, after the dialog submission, we could give the option to trigger iframe reload like with:
`document.getElementsByTagName('iframe')[0].contentWindow.location.reload();` 
instead of reloading the document node only?

SPA frameworks like React, Angular, Vue do not play well with direct DOM manipulations.
Reloading the whole iframe may result in a screen blink, but it may be the simplest and robust solution, that should work for all SPA frameworks without any custom code.

**Reply by Harikarthyk** [reply]
@michalcukierman Okay, thanks. I could not figure out any function related to that submit button in the dialog. Where can we find that or add the custom function for the "Submit" button in the Edit dialogue box?

**Reply by michalcukierman** [reply]
@paweljozwicki is it possible at the moment?

**Reply by paweljozwicki** [reply]
I'll start with some explanation after investigating the problem. 
As @michalcukierman mentioned SPA frameworks do not play well with direct DOM manipulations and the problem occurs not only on submitting dialog changes, but also adding and removing components. Each of those actions uses `EditModeStore` from `websight-pages-editor-view` to manipulate DOM, but it's not prepared to support custom actions. Now we are working on changes for Page Editor to support adding custom plugins, but until then I think it won't be possible to do something with that. 

We were not yet focusing on supporting SPAs. I'm still looking for some solution, but I'm afraid it won't work well without waiting for changes in WebSight CMS.

**Comment by paweljozwicki** [comment]
Hi, just to summarise:
we will deliver some solutions that should help with your issue. 
 
First of all, we will provide a possibility to define if component changes should re-render the whole iframe. In the case of SPA, all the components would have to use this mechanism, but it should help you because refreshing the iframe would rebuild the whole Vue app. 

  We will also provide a possibility to extend Page Editor with custom actions, but for now, I won’t share any details, because we have to design it first.

**Reply by Harikarthyk** [reply]
Hi, @paweljozwicki 

Any updates on this?

**Reply by paweljozwicki** [reply]
Hi, we have it implemented, but not been documented yet. 

What you have to do to extend the page editor is to add a new WebFragment:

``` java
package pl.ds.howlite;

import org.osgi.service.component.annotations.Component;
import pl.ds.websight.fragments.registry.WebFragment;

@Component
public class RefreshVueStateExtension implements WebFragment {

  @Override
  public String getKey() {
    return "websight.editor.spi.extension";
  }

  @Override
  public String getFragment() {
    return "/libs/howlite/author/editor/extensions/RefreshVueStateExtension.js";
  }

  @Override
  public int getRanking() {
    return 200;
  }
}
```

And it should point to the js file that does iframe refreshing.

``` JS
export default {
  init: (editor) => {
    editor.addEventListener('component-dom-updated', () => {
      editor.componentTree.get().parentDocument.location.reload();
    });
  }
}
``` 

To use WebFragment you will have to add this dependency to your project:

``` xml
<dependency>
    <groupId>pl.ds.websight</groupId>
    <artifactId>websight-fragments-registry</artifactId>
    <version>1.0.3</version>
    <scope>provided</scope>
</dependency>
```

More details should be available soon in the documentation, but I hope it will help you meanwhile.

**Reply by paweljozwicki** [reply]
Hi, [here is the documentation](https://www.websight.io/docs/developers/development/page-editor/extending-editor/) explaining the solution I sent earlier.

**Reply by Harikarthyk** [reply]
Ok, thanks for sharing.

We will incorporate them into ours and update you.
